compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3482,3 +3482,86 @@ fn make_thin_self_ptr<'tcx>(
         ..tcx.layout_of(ty::ParamEnv::reveal_all().and(unit_ptr_ty)).unwrap()
     }
 }
+
+/// Determines if this type permits "raw" initialization by just transmuting some
+/// uninitialized memory into an instance of `T`.
```

Comment:
"raw_init" was picked as a name to generalize over uninit and zero-init. So once this is uninit-only I think it should be renamed to "might_permit_uninit" or so.

But I would also hold off on that part of the change for now. Old crossbeam will not like it.

Reply:
So we *should* be looking inside arrays for uninit checks, but not for zeroing checks? Since old crosssbeam makes what is effectively a zeroed `[fn(); N]`

Reply:
I think this PR should be about uninit only, yes. There we have a plan that avoids LLVM UB.

For crossbeam, no such plan can exist. Old crossbeam unambiguously has LLVM UB and we cannot work around this since `zeroed` is pretty clear about what the memory it returns will look like. So that is a separate discussion IMO. (FWIW we totally could return 0x01-filled memory when we *know* the code caused UB, since we can do anything then. Not sure if we want to.^^)

Reply:
There are a *few* breakages from old crossbeam making uninit values, but from the crater run, it's a few (2 github regressions and `anterofit 0.2.0` and `constellation 0.2.0`), both having minimal download count. So yeah, ignoring zeroed failures inside arrays looks like it'll cut down on nearly all the crossbeam issues.